File: lib/g3/tracker/fact_search.ex
defmodule G3.Tracker.FactSearch do
  @moduledoc false

  @default_limit 5
  @k1 1.2
  @b 0.75

  def bm25(query, facts_or_snapshot, opts \\ []) when is_binary(query) do
    facts = facts_list(facts_or_snapshot)
    limit = Keyword.get(opts, :limit, @default_limit)
    query_tokens = query |> tokenize() |> Enum.uniq()

    documents = Enum.map(facts, &build_document/1)
    document_count = max(length(documents), 1)
    average_length = average_document_length(documents)
    document_frequencies = document_frequencies(documents, query_tokens)

    documents
    |> Enum.map(fn document ->
      score =
        bm25_score(document, query_tokens, document_frequencies, document_count, average_length)

      result(document.fact, score, best_snippet(document.fact, query_tokens))
    end)
    |> Enum.filter(&(&1["score"] > 0.0))
    |> Enum.sort_by(&{-&1["score"], &1["title"]})
    |> Enum.take(limit)
  end

  def grep(pattern, facts_or_snapshot, opts \\ []) when is_binary(pattern) do
    facts = facts_list(facts_or_snapshot)
    limit = Keyword.get(opts, :limit, @default_limit)
    regex = compile_pattern(pattern)

    facts
    |> Enum.reduce([], fn fact, acc ->
      case first_match(fact, regex) do
        nil ->
          acc

        %{field: field, snippet: snippet, matches: matches} ->
          [grep_result(fact, field, snippet, matches) | acc]
      end
    end)
    |> Enum.reverse()
    |> Enum.take(limit)
  end

  defp facts_list(%{"facts" => facts}) when is_list(facts), do: facts
  defp facts_list(facts) when is_list(facts), do: facts
  defp facts_list(_other), do: []

  defp build_document(fact) do
    tokens =
      weighted_tokens(fact["title"], 4) ++
        weighted_tokens(fact["summary"], 2) ++
        weighted_tokens(fact["details"], 3) ++
        weighted_tokens(fact["project_title"], 2)

    %{
      fact: fact,
      tokens: tokens,
      token_frequencies: Enum.frequencies(tokens),
      length: max(length(tokens), 1)
    }
  end

  defp weighted_tokens(nil, _weight), do: []

  defp weighted_tokens(text, weight) when is_binary(text) and weight > 0 do
    text
    |> tokenize()
    |> Enum.flat_map(fn token -> List.duplicate(token, weight) end)
  end

  defp tokenize(text) when is_binary(text) do
    text
    |> String.downcase()
    |> String.split(~r/[^a-z0-9]+/u, trim: true)
  end

  defp tokenize(_text), do: []

  defp average_document_length([]), do: 1.0

  defp average_document_length(documents) do
    documents
    |> Enum.map(& &1.length)
    |> Enum.sum()
    |> Kernel./(length(documents))
  end

  defp document_frequencies(documents, query_tokens) do
    Enum.reduce(query_tokens, %{}, fn token, acc ->
      frequency =
        Enum.count(documents, fn document ->
          Map.has_key?(document.token_frequencies, token)
        end)

      Map.put(acc, token, frequency)
    end)
  end

  defp bm25_score(document, query_tokens, document_frequencies, document_count, average_length) do
    Enum.reduce(query_tokens, 0.0, fn token, score ->
      term_frequency = Map.get(document.token_frequencies, token, 0)

      if term_frequency == 0 do
        score
      else
        document_frequency = Map.get(document_frequencies, token, 0)

        idf =
          :math.log(1 + (document_count - document_frequency + 0.5) / (document_frequency + 0.5))

        numerator = term_frequency * (@k1 + 1)
        denominator = term_frequency + @k1 * (1 - @b + @b * document.length / average_length)

        score + idf * numerator / denominator
      end
    end)
  end

  defp result(fact, score, snippet) do
    %{
      "id" => fact["id"],
      "title" => fact["title"],
      "project_title" => fact["project_title"],
      "score" => Float.round(score, 4),
      "snippet" => snippet
    }
  end

  defp best_snippet(fact, query_tokens) do
    text =
      Enum.find_value(
        [fact["details"], fact["summary"], fact["title"], fact["project_title"]],
        fn value ->
          if is_binary(value) and
               Enum.any?(query_tokens, &String.contains?(String.downcase(value), &1)) do
            value
          end
        end
      ) || fact["details"] || fact["summary"] || fact["title"] || fact["project_title"] || ""

    snippet(text)
  end

  defp snippet(text) when is_binary(text) do
    if String.length(text) > 180 do
      String.slice(text, 0, 177) <> "..."
    else
      text
    end
  end

  defp compile_pattern(pattern) do
    case Regex.compile(pattern, "iu") do
      {:ok, regex} ->
        regex

      {:error, _reason} ->
        {:ok, regex} = Regex.compile(Regex.escape(pattern), "iu")
        regex
    end
  end

  defp first_match(fact, regex) do
    Enum.find_value(
      [
        {"title", fact["title"]},
        {"summary", fact["summary"]},
        {"details", fact["details"]},
        {"project_title", fact["project_title"]}
      ],
      fn
        {_field, nil} ->
          nil

        {field, value} ->
          matches = Regex.scan(regex, value) |> length()

          if matches > 0 do
            %{field: field, snippet: snippet(value), matches: matches}
          end
      end
    )
  end

  defp grep_result(fact, field, snippet, matches) do
    %{
      "id" => fact["id"],
      "title" => fact["title"],
      "project_title" => fact["project_title"],
      "field" => field,
      "matches" => matches,
      "snippet" => snippet
    }
  end
end

File: native/g3_core/src/fact_search.rs
use std::collections::{HashMap, HashSet};

use regex::Regex;
use serde::Serialize;

use crate::models::TrackerItem;

const DEFAULT_LIMIT: usize = 5;
const K1: f64 = 1.2;
const B: f64 = 0.75;

#[derive(Debug, Clone, Serialize)]
pub struct SearchResult {
    pub id: String,
    pub title: Option<String>,
    pub project_title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub score: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub field: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub matches: Option<usize>,
    pub snippet: String,
}

pub fn bm25(query: &str, facts: &[TrackerItem], limit: usize) -> Vec<SearchResult> {
    let limit = if limit == 0 { DEFAULT_LIMIT } else { limit };
    let query_tokens = tokenize(query).into_iter().collect::<HashSet<_>>();
    if query_tokens.is_empty() {
        return Vec::new();
    }

    let documents = facts.iter().map(Document::new).collect::<Vec<_>>();
    let document_count = documents.len().max(1) as f64;
    let average_length = average_document_length(&documents);
    let frequencies = document_frequencies(&documents, &query_tokens);

    let mut scored = documents
        .into_iter()
        .filter_map(|document| {
            let score = bm25_score(
                &document,
                &query_tokens,
                &frequencies,
                document_count,
                average_length,
            );

            if score > 0.0 {
                Some(SearchResult {
                    id: document.fact.id.clone(),
                    title: document.fact.title.clone(),
                    project_title: document.fact.project_title.clone(),
                    score: Some((score * 10_000.0).round() / 10_000.0),
                    field: None,
                    matches: None,
                    snippet: best_snippet(document.fact, &query_tokens),
                })
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    scored.sort_by(|left, right| {
        right
            .score
            .partial_cmp(&left.score)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| left.title.cmp(&right.title))
    });
    scored.truncate(limit);
    scored
}

pub fn grep(pattern: &str, facts: &[TrackerItem], limit: usize) -> Vec<SearchResult> {
    let limit = if limit == 0 { DEFAULT_LIMIT } else { limit };
    let regex = Regex::new(pattern)
        .or_else(|_| Regex::new(&regex::escape(pattern)))
        .ok();

    let Some(regex) = regex else {
        return Vec::new();
    };

    facts
        .iter()
        .filter_map(|fact| {
            for (field, value) in [
                ("title", fact.title.as_deref()),
                ("summary", fact.summary.as_deref()),
                ("details", fact.details.as_deref()),
                ("project_title", fact.project_title.as_deref()),
            ] {
                let Some(value) = value else {
                    continue;
                };
                let matches = regex.find_iter(value).count();
                if matches > 0 {
                    return Some(SearchResult {
                        id: fact.id.clone(),
                        title: fact.title.clone(),
                        project_title: fact.project_title.clone(),
                        score: None,
                        field: Some(field.to_string()),
                        matches: Some(matches),
                        snippet: snippet(value),
                    });
                }
            }
            None
        })
        .take(limit)
        .collect()
}

struct Document<'a> {
    fact: &'a TrackerItem,
    tokens: Vec<String>,
    frequencies: HashMap<String, usize>,
}

impl<'a> Document<'a> {
    fn new(fact: &'a TrackerItem) -> Self {
        let tokens = weighted_tokens(fact.title.as_deref(), 4)
            .into_iter()
            .chain(weighted_tokens(fact.summary.as_deref(), 2))
            .chain(weighted_tokens(fact.details.as_deref(), 3))
            .chain(weighted_tokens(fact.project_title.as_deref(), 2))
            .collect::<Vec<_>>();

        let mut frequencies = HashMap::new();
        for token in &tokens {
            *frequencies.entry(token.clone()).or_insert(0) += 1;
        }

        Self {
            fact,
            tokens,
            frequencies,
        }
    }
}

fn weighted_tokens(text: Option<&str>, weight: usize) -> Vec<String> {
    let Some(text) = text else {
        return Vec::new();
    };

    tokenize(text)
        .into_iter()
        .flat_map(|token| std::iter::repeat(token).take(weight))
        .collect()
}

fn tokenize(text: &str) -> Vec<String> {
    text.to_lowercase()
        .split(|ch: char| !ch.is_ascii_alphanumeric())
        .filter(|token| !token.is_empty())
        .map(ToString::to_string)
        .collect()
}

fn average_document_length(documents: &[Document<'_>]) -> f64 {
    if documents.is_empty() {
        1.0
    } else {
        documents
            .iter()
            .map(|document| document.tokens.len() as f64)
            .sum::<f64>()
            / documents.len() as f64
    }
}

fn document_frequencies(
    documents: &[Document<'_>],
    query_tokens: &HashSet<String>,
) -> HashMap<String, usize> {
    query_tokens
        .iter()
        .map(|token| {
            let frequency = documents
                .iter()
                .filter(|document| document.frequencies.contains_key(token))
                .count();
            (token.clone(), frequency)
        })
        .collect()
}

fn bm25_score(
    document: &Document<'_>,
    query_tokens: &HashSet<String>,
    document_frequencies: &HashMap<String, usize>,
    document_count: f64,
    average_length: f64,
) -> f64 {
    query_tokens.iter().fold(0.0, |score, token| {
        let term_frequency = *document.frequencies.get(token).unwrap_or(&0) as f64;
        if term_frequency == 0.0 {
            return score;
        }

        let document_frequency = *document_frequencies.get(token).unwrap_or(&0) as f64;
        let idf =
            (1.0 + (document_count - document_frequency + 0.5) / (document_frequency + 0.5)).ln();
        let numerator = term_frequency * (K1 + 1.0);
        let denominator =
            term_frequency + K1 * (1.0 - B + B * document.tokens.len() as f64 / average_length);
        score + idf * numerator / denominator
    })
}

fn best_snippet(fact: &TrackerItem, query_tokens: &HashSet<String>) -> String {
    for value in [
        fact.details.as_deref(),
        fact.summary.as_deref(),
        fact.title.as_deref(),
        fact.project_title.as_deref(),
    ] {
        if let Some(value) = value {
            let lower = value.to_lowercase();
            if query_tokens.iter().any(|token| lower.contains(token)) {
                return snippet(value);
            }
        }
    }

    snippet(
        fact.details
            .as_deref()
            .or(fact.summary.as_deref())
            .or(fact.title.as_deref())
            .or(fact.project_title.as_deref())
            .unwrap_or_default(),
    )
}

fn snippet(text: &str) -> String {
    if text.chars().count() > 180 {
        format!("{}...", text.chars().take(177).collect::<String>())
    } else {
        text.to_string()
    }
}
